### PR TITLE
[inductor] Specialize on unguarded alignment of example inputs

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124336
* __->__ #123319

When inductor generates triton code, the triton code can either assume that the inputs given to it are aligned or unaligned. If they are aligned, triton can use more efficient instructions (like vectorized loads or tensor cores). However, if we generate "aligned" code and pass in unaligned inputs, the triton code will error out; to fix this, we clone unaligned inputs that are passed to triton kernels that expect aligned inputs. This can lead to excessive clones if we have inputs that are not expected to be aligned.

In this PR, we use the example input to decide whether the generated triton code should assume alignment or not. If the example input is aligned, then we will generate triton code that assumes alignment; if at runtime we receive an unaligned input, we'll make a clone. Meanwhile, if the example input is not aligned, the generated triton code will not assume inputs are aligned and we won't ever need to clone.

Note that the alignment of the inputs is not guarded on; we found that adding guards on tensor offsets (a) was slow in cases where we do a lot of comparisons on tensor offsets, and (b) led to a lot of recompilations.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/compile_fx.py [hunk 1]
```diff
@@ -799,17 +798,19 @@ def get_input_idxs_to_check(
     inputs: Union[List[torch.Tensor], Sequence[int]],
     static_input_idxs: Sequence[int],
 ) -> Sequence[int]:
-    def is_aligned(storage_offset, dtype):
+    def is_aligned(input):
```
Comment (bdhirsh) [inline torch/_inductor/compile_fx.py:801]: nit: this logic is duped in `utils.py`, can probably just share it there?

Comment (davidberard98) [reply torch/_inductor/compile_fx.py]: thanks, I think you're right that the previous logic was incorrect

File: torch/_inductor/scheduler.py [hunk 2]
```diff
@@ -2497,7 +2497,8 @@ def codegen(self):
 
     def is_unaligned_buffer(self, buf_name):
         if buf_name in V.graph.graph_inputs:
-            return not config.assume_aligned_inputs
+            # See Note: [Input Alignment handling in Inductor]
+            return buf_name not in V.graph.aligned_inputs
```
Comment (eellison) [inline torch/_inductor/scheduler.py:2501]: I think you'll need to rebase past https://github.com/pytorch/pytorch/pull/124030/files#diff-34f851dde2ac5502be643489261bb0115d0cd873fac5367831f9850b471b3279R68

File: torch/_inductor/graph.py [hunk 3]
```diff
@@ -856,6 +860,21 @@ def placeholder(self, target: str, args, kwargs):
...
+        # But if we codegen assuming aligned inputs and then get unaligned
+        # inputs at runtime, then we are forced to clone - which is bad for
+        # both perf and memory usage.
+        #
+        # One option would be to guard on storage_offset%ALIGNMENT, and then
+        # codegen based on this. But storage_offset guards turned out to be
+        # expensive and cause recompiles; Instead, we're generating code
+        # based on the alignment of the example input without guarding.
+        if should_assume_input_aligned(example):
+            self.aligned_inputs.add(target)
```
Comment (eellison) [inline torch/_inductor/graph.py:877]:
```
import torch

@torch.compile()
def foo(x):
    return x.add_(2)

y = torch.zeros([4], device="cuda")
z = y[1:]
foo(z)
print(z)
```

Let's also not assume alignment on mutated inputs to fix the above ^^ issue

**Review by eellison — APPROVED** [review]
looks good to me ! let's fix one bug while we're at it.

File: torch/_inductor/graph.py (hunk 3 above)
Comment (davidberard98) [reply torch/_inductor/graph.py:877]: what's the issue here? I ran this and it seems to return `[2., 2., 2.]` as expected

(also - didn't realize inductor had to deal with mutations? I thought aot autograd / functionalization would deal with this?)

Comment (davidberard98) [reply torch/_inductor/graph.py:877]: Also for this, I think we'd need to propagate this info to compile_fx.py, which isn't being done right now. I think this is somewhat non-trivial:

- right now, graph.py aligned input handling is only used to do codegen
- separately, in compile_fx.py we also check should_assume_input_aligned to see whether we should do a clone
- it seems like if we want to rely on graph.py artifacts in compile_fx.py, this is a bit difficult because we cache the resulting artifacts sometimes (and in the cached case I don't think we have access to self.aligned_inputs in compile_fx.py?)

Comment (eellison) [reply torch/_inductor/graph.py:877]: Try running this :
import torch

@torch.compile()
def foo(x):
    return x.add_(2)

y = torch.zeros([16], device="cuda")
# specialize on aligned
foo(y[8:12])

# re-run with unaligned will clone input
z = y[1:5]
print(z)
```

Yes - for inference (and maybe training? not sure) we now add back in the copy_ kernels into the graph instead of running them in aot autograd

File: torch/_inductor/utils.py [hunk 4]
```diff
@@ -1558,3 +1560,16 @@ def use_scatter_fallback(
         or (reduction_type == reduce_ty and self_dtype in {torch.bool, torch.int64})
         or torch.are_deterministic_algorithms_enabled()
     )
+
+
+def tensor_is_aligned(tensor: torch.Tensor):
+    return (tensor.storage_offset() * get_dtype_size(tensor.dtype)) % ALIGNMENT == 0
+
```
Comment (eellison) [inline torch/_inductor/utils.py:1567]: Just read your description of the second PR - we should evaluate this expression with suppressed guards. We don't want to add further compile specialization but just guide our runtime behavior.

Comment (eellison) [reply torch/_inductor/utils.py:1567]: `with TracingContext.get().fake_mode.shape_env.suppress_guards():`

File: torch/_inductor/graph.py (hunk 3 above)
Comment (eellison) [reply torch/_inductor/graph.py:877]: Hmm maybe we can do this in a separate PR. not sure if this will lead to perf decrease.

Comment (bdhirsh) [reply torch/_inductor/graph.py:876]: > we now add back in the copy_ kernels into the graph instead of running them in aot autograd

Yup we now do this for training too (the tldr is that as long as the input mutation doesn't mutate the grad_fn of the input, we try to keep it in the graph)

**Review by eellison — APPROVED** [review]
looks great! 

Not imminent, but maybe a couple things we could consider doing in the future is :

1) Figure out a solution for aligning tensors which are mutated.. In this case, I think it's possible we'd want to specialize on if the inputs are views and re-compiling. Since we don't have the option of runtime cloning for alignment. But anyways let's defer this. 

2) More aggressively assume alignment on tensors which are used in mm/conv.. I would be kind of surprised if we had a much of indexed, unaligned tensors going into these kernels. 

Anyways these are just thoughts I don't know if either is that important.

cc @desertfire for aot inductor, we may need to do similar runtime clones
